source/common/network/buffered_io_socket_handle_impl.h

```diff
@@ -0,0 +1,145 @@
+                                  uint32_t self_port, RecvMsgOutput& output) override;
+  Api::IoCallUint64Result recvmmsg(RawSliceArrays& slices, uint32_t self_port,
+                                   RecvMsgOutput& output) override;
+  Api::IoCallUint64Result recv(void* buffer, size_t length, int flags) override;
+  bool supportsMmsg() const override;
+  bool supportsUdpGro() const override;
+  Api::SysCallIntResult bind(Address::InstanceConstSharedPtr address) override;
+  Api::SysCallIntResult listen(int backlog) override;
+  IoHandlePtr accept(struct sockaddr* addr, socklen_t* addrlen) override;
+  Api::SysCallIntResult connect(Address::InstanceConstSharedPtr address) override;
+  Api::SysCallIntResult setOption(int level, int optname, const void* optval,
+                                  socklen_t optlen) override;
+  Api::SysCallIntResult getOption(int level, int optname, void* optval, socklen_t* optlen) override;
+  Api::SysCallIntResult setBlocking(bool blocking) override;
+  absl::optional<int> domain() override;
+  Address::InstanceConstSharedPtr localAddress() override;
+  Address::InstanceConstSharedPtr peerAddress() override;
+
+  void initializeFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                           Event::FileTriggerType trigger, uint32_t events) override;
+  IoHandlePtr duplicate() override;
+  void activateFileEvents(uint32_t events) override;
+  void enableFileEvents(uint32_t events) override;
+  void resetFileEvents() override;
+
+  Api::SysCallIntResult shutdown(int how) override;
+  absl::optional<std::chrono::milliseconds> lastRoundTripTime() override { return absl::nullopt; }
+
+  Buffer::WatermarkBuffer& getBufferForTest() { return pending_received_data_; }
+
+  void setWritablePeer(WritablePeer* writable_peer) {
+    // Swapping writable peer is undefined behavior.
+    ASSERT(!writable_peer_);
+    ASSERT(!write_shutdown_);
+    writable_peer_ = writable_peer;
+  }
+
+  // WritablePeer
+  void setWriteEnd() override { read_end_stream_ = true; }
+  bool isWriteEndSet() override { return read_end_stream_; }
+  void maybeSetNewData() override {
+    ENVOY_LOG(trace, "{} on socket {}", __FUNCTION__, static_cast<void*>(this));
+    if (user_file_event_) {
+      user_file_event_->activate(Event::FileReadyType::Read);
+    }
+  }
+  void onPeerDestroy() override {
+    writable_peer_ = nullptr;
+    write_shutdown_ = true;
+  }
+  void onPeerBufferWritable() override {
+    if (user_file_event_) {
+      user_file_event_->activate(Event::FileReadyType::Write);
+    }
+  }
+  bool isWritable() const override { return !isOverHighWatermark(); }
+  bool isPeerWritable() const override {
+    return writable_peer_ != nullptr && !writable_peer_->isWriteEndSet() &&
+           writable_peer_->isWritable();
+  }
+  Buffer::Instance* getWriteBuffer() override { return &pending_received_data_; }
+
+  // ReadableSource
```

Comment:
Is this interface only for use in tests?  I can't seem to find meaningful uses of these methods even in tests.  The relationship between isWritable and isOverHighWatermark seems a bit surprising.

Reply:
As explained in https://github.com/envoyproxy/envoy/pull/13418#discussion_r515403987

Right, this ReadableSource are cohesive methods but are not deserved a name. Will delete this interface